Allow less restrictive tensor checks

BrianHarrisonAMD opened the pull request:
Based off the changes proposed in #3763, but this is a general check that will allow skipping to the next set of strides when dim == 1.


Note: still need to add more tests to cover these conditions, but opening the PR now to get feedback on changes, and run the CI.

Cemberk committed:
Update tensor.cpp

Cemberk committed:
Update tensor.cpp

Cemberk committed:
Update tensor.cpp

BrianHarrisonAMD committed:
fix formatting

BrianHarrisonAMD committed:
Remove redundant check

BrianHarrisonAMD committed:
Allow less restrictive tensor checks

BrianHarrisonAMD committed:
Formatting fix

BrianHarrisonAMD commented on src/tensor.cpp:
```diff
@@ -674,23 +674,46 @@ bool TensorDescriptor::IsPossibleLayout(const std::string& storage_layout,
     {
-        if(layout_strides[i] < layout_strides[i + 1])
+        // potential violation detected
+        if(layout_strides[current_index] < layout_strides[next_index])
```
I think it should also skip if the current_index has a dim of 1, and generally should be modified to skip checks when dim == 1.

BrianHarrisonAMD committed:
Fix tests to take into account dim 1 logic, and add additional checks

BrianHarrisonAMD committed:
Fix checking logic to take into account skipping dim == 1 for current

BrianHarrisonAMD commented on test/gtest/unit_TensorDescriptor.cpp:
```diff
@@ -198,31 +198,60 @@ class TestPossibleLayout4D5D : public ::testing::TestWithParam<TestCasePossibleL
             TestCase{{miopenHalf, {2, 2, 2, 2}, {1000, 1, 100, 10}}, {"NHWC"}},
             TestCase{{miopenHalf, {2, 2, 2, 2}, {1, 1000, 100, 10}}, {"CHWN"}},
             TestCase{{miopenHalf, {2, 2, 2, 2}, {1, 1, 1, 1}}, {"NCHW", "NHWC", "CHWN"}},
-            TestCase{{miopenHalf, {1, 1, 1, 1, 1}, {10000, 1000, 100, 10, 1}}, {"NCDHW"}},
-            TestCase{{miopenHalf, {1, 1, 1, 1, 1}, {10000, 1, 1000, 100, 10}}, {"NDHWC"}},
-            TestCase{{miopenHalf, {1, 1, 1, 1, 1}, {1, 1, 1, 1, 1}}, {"NCDHW", "NDHWC"}},
+            TestCase{{miopenHalf, {2, 2, 2, 2, 2}, {10000, 1000, 100, 10, 1}}, {"NCDHW"}},
+            TestCase{{miopenHalf, {2, 2, 2, 2, 2}, {10000, 1, 1000, 100, 10}}, {"NDHWC"}},
+            TestCase{{miopenHalf, {2, 2, 2, 2, 2}, {1, 1, 1, 1, 1}}, {"NCDHW", "NDHWC"}},
             TestCase{{miopenHalf, {2, 2, 2, 2, 2}, {10000, 1000, 100, 10, 1}}, {"NCDHW"}},
             TestCase{{miopenHalf, {2, 2, 2, 2, 2}, {10000, 1, 1000, 100, 10}}, {"NDHWC"}},
             TestCase{{miopenHalf, {2, 2, 2, 2, 2}, {1, 1, 1, 1, 1}}, {"NCDHW", "NDHWC"}},
 
-            TestCase{{miopenHalf, miopenTensorNCHW, {1, 1, 1, 1}, {1000, 100, 10, 1}}, {"NCHW"}},
-            TestCase{{miopenHalf, miopenTensorNHWC, {1, 1, 1, 1}, {1000, 1, 100, 10}}, {"NHWC"}},
-            TestCase{{miopenHalf, miopenTensorCHWN, {1, 1, 1, 1}, {1, 1000, 100, 10}}, {"CHWN"}},
             TestCase{{miopenHalf, miopenTensorNCHW, {2, 2, 2, 2}, {1000, 100, 10, 1}}, {"NCHW"}},
             TestCase{{miopenHalf, miopenTensorNHWC, {2, 2, 2, 2}, {1000, 1, 100, 10}}, {"NHWC"}},
             TestCase{{miopenHalf, miopenTensorCHWN, {2, 2, 2, 2}, {1, 1000, 100, 10}}, {"CHWN"}},
-            TestCase{{miopenHalf, miopenTensorNCDHW, {1, 1, 1, 1, 1}, {10000, 1000, 100, 10, 1}}, {"NCDHW"}},
-            TestCase{{miopenHalf, miopenTensorNDHWC, {1, 1, 1, 1, 1}, {10000, 1, 1000, 100, 10}}, {"NDHWC"}},
+            TestCase{{miopenHalf, miopenTensorNCHW, {2, 2, 2, 2}, {1000, 100, 10, 1}}, {"NCHW"}},
+            TestCase{{miopenHalf, miopenTensorNHWC, {2, 2, 2, 2}, {1000, 1, 100, 10}}, {"NHWC"}},
+            TestCase{{miopenHalf, miopenTensorCHWN, {2, 2, 2, 2}, {1, 1000, 100, 10}}, {"CHWN"}},
```
All existing checks that had dims of 1, now need to be changed to dims > 1. 

Added extra checks to verify more relaxed layout logic as well.

randyspauldingamd replied:
I think we already had all of the 2's. I marked them so you can double-check.

BrianHarrisonAMD replied:
Yea very clear now that I am looking at it again that these are duped.
Ill remove, and only keep the new ones.

BrianHarrisonAMD committed:
Remove duplicates

BrianHarrisonAMD committed:
Swap to a filtering technique that will make checks less confusing

BrianHarrisonAMD committed:
Update comment to be more clear

BrianHarrisonAMD committed:
Formatting

BrianHarrisonAMD committed:
Tidy fix

randyspauldingamd commented on test/gtest/unit_TensorDescriptor.cpp:
```diff
@@ -198,31 +198,58 @@ class TestPossibleLayout4D5D : public ::testing::TestWithParam<TestCasePossibleL
-            TestCase{{miopenHalf, {1, 1, 1, 1, 1}, {1, 1, 1, 1, 1}}, {"NCDHW", "NDHWC"}},
+            TestCase{{miopenHalf, {2, 2, 2, 2, 2}, {10000, 1000, 100, 10, 1}}, {"NCDHW"}},
+            TestCase{{miopenHalf, {2, 2, 2, 2, 2}, {10000, 1, 1000, 100, 10}}, {"NDHWC"}},
+            TestCase{{miopenHalf, {2, 2, 2, 2, 2}, {1, 1, 1, 1, 1}}, {"NCDHW", "NDHWC"}},
```
Duplicate of line 214. At first glance it doesn't look like it should be valid, although having this test case simply marks it as an intentional way users can save memory for tensors where all values are the same. I don't see a reason to disallow this. We should add the 4D case as well.

BrianHarrisonAMD replied:
I think we have the 4D case isn't it same as:
` TestCase{{miopenHalf, {2, 2, 2, 2}, {1, 1, 1, 1}}, {"NCHW", "NHWC", "CHWN"}},` ?

BrianHarrisonAMD committed:
Merge branch 'develop' into bharriso/allow-less-restrictive-tensor-checks

randyspauldingamd commented on test/gtest/unit_TensorDescriptor.cpp:
```diff
@@ -198,31 +198,58 @@ class TestPossibleLayout4D5D : public ::testing::TestWithParam<TestCasePossibleL
-            TestCase{{miopenHalf, {1, 1, 1, 1}, {1, 1000, 100, 10}}, {"CHWN"}},
+            TestCase{{miopenHalf, {2, 2, 2, 2}, {1000, 100, 10, 1}}, {"NCHW"}},
+            TestCase{{miopenHalf, {2, 2, 2, 2}, {1000, 1, 100, 10}}, {"NHWC"}},
+            TestCase{{miopenHalf, {2, 2, 2, 2}, {1, 1000, 100, 10}}, {"CHWN"}},
```
Duplicates of 205-207

randyspauldingamd commented on test/gtest/unit_TensorDescriptor.cpp:
```diff
@@ -198,31 +198,58 @@ class TestPossibleLayout4D5D : public ::testing::TestWithParam<TestCasePossibleL
-            TestCase{{miopenHalf, {1, 1, 1, 1, 1}, {10000, 1, 1000, 100, 10}}, {"NDHWC"}},
-            TestCase{{miopenHalf, {1, 1, 1, 1, 1}, {1, 1, 1, 1, 1}}, {"NCDHW", "NDHWC"}},
+            TestCase{{miopenHalf, {2, 2, 2, 2, 2}, {10000, 1000, 100, 10, 1}}, {"NCDHW"}},
+            TestCase{{miopenHalf, {2, 2, 2, 2, 2}, {10000, 1, 1000, 100, 10}}, {"NDHWC"}},
```
Duplicates of 212-213

randyspauldingamd commented on test/gtest/unit_TensorDescriptor.cpp:
```diff
@@ -198,31 +198,58 @@ class TestPossibleLayout4D5D : public ::testing::TestWithParam<TestCasePossibleL
-            TestCase{{miopenHalf, miopenTensorNDHWC, {1, 1, 1, 1, 1}, {10000, 1, 1000, 100, 10}}, {"NDHWC"}},
+            TestCase{{miopenHalf, miopenTensorNCHW, {2, 2, 2, 2}, {1000, 100, 10, 1}}, {"NCHW"}},
+            TestCase{{miopenHalf, miopenTensorNHWC, {2, 2, 2, 2}, {1000, 1, 100, 10}}, {"NHWC"}},
+            TestCase{{miopenHalf, miopenTensorCHWN, {2, 2, 2, 2}, {1, 1000, 100, 10}}, {"CHWN"}},
```
Duplicates of 216-218

BrianHarrisonAMD replied:
Oh good point, I didn't realize we already had these exact cases with non-1 dims hehe.

Ill remove the dupes and only keep the new checks.

randyspauldingamd commented on test/gtest/unit_TensorDescriptor.cpp:
```diff
@@ -198,31 +198,58 @@ class TestPossibleLayout4D5D : public ::testing::TestWithParam<TestCasePossibleL
             TestCase{{miopenHalf, miopenTensorNCDHW, {2, 2, 2, 2, 2}, {10000, 1000, 100, 10, 1}}, {"NCDHW"}},
             TestCase{{miopenHalf, miopenTensorNDHWC, {2, 2, 2, 2, 2}, {10000, 1, 1000, 100, 10}}, {"NDHWC"}},
+
+            // Test cases where invalid strides with dim of 1 are allowed.
+            // (When dim is 1 we can safely ignore stride values.)
+            // Note: Some layouts are equivalent when dim == 1 (NCHW == CHWN when n = 1 for example)
```
Suggested rewording (feel free to continue tweaking as you see fit):
```
// Stride is ignored when the corresponding dimension length is 1. As a result,
// multiple layouts have identical memory footprints. For example, NCHW == CHWN for n = 1.
```

randyspauldingamd reviewed (commented):
I would greatly prefer strict checking to help users catch mistakes early on, but I don't think it's necessary to impose it upon them. Forge onward!

BrianHarrisonAMD committed:
code review feedback

BrianHarrisonAMD committed:
Fix test cases where GraphAPI assumes NCHW layout instead of NHWC

BrianHarrisonAMD commented on test/gtest/graphapi_conv_bias_res_add_activ_fwd.cpp:
```diff
@@ -55,9 +55,9 @@ static std::vector<Conv3DTestCase> ConvTestConfigs()
-            {1, 1, 1, 1, {1, 4, 4}, {2, 2, 2}, {1, 1, 1}, {1, 1, 1}, {1, 1, 1}, miopenConvolution},
-            {1, 1, 1, 1, {8, 8, 8}, {2, 2, 2}, {0, 0, 0}, {1, 1, 1}, {1, 1, 1}, miopenConvolution},
-            {1, 1, 1, 1, {8, 8, 8}, {2, 2, 2}, {0, 0, 0}, {2, 2, 2}, {1, 1, 1}, miopenConvolution},
+            {1, 1, 2, 1, {1, 4, 4}, {2, 2, 2}, {1, 1, 1}, {1, 1, 1}, {1, 1, 1}, miopenConvolution},
+            {1, 1, 2, 1, {8, 8, 8}, {2, 2, 2}, {0, 0, 0}, {1, 1, 1}, {1, 1, 1}, miopenConvolution},
+            {1, 1, 2, 1, {8, 8, 8}, {2, 2, 2}, {0, 0, 0}, {2, 2, 2}, {1, 1, 1}, miopenConvolution},
```
Edit: Looked into this more, and think it might be something odd with the graphAPI path of how the layout gets decided for fusions that leads to it thinking the tensors are NCHW when fusion only supports NHWC.

Going to run the remaining tests to see if there is anything else having issues or if this was isolated to this graphAPI test that we are deprecating and removing eventually.

BrianHarrisonAMD committed:
Fix replacing calculated layouts with possible layouts for convs

randyspauldingamd reviewed (approved):
I guess they aren't going to supply more information in response to Evgenii's questions, so this LGTM unless you want to ping them again.

However, this has made it more clear how the string layouts are a limiting design flaw. What we need for more generic use cases such as this is a set of enums so the solvers can quickly determine whether the memory layout is appropriate for their kernels by checking flag(s). That of course would require a very large refactor. We could mimic it quite easily using the existing API though with little added compute expense, by simply including all valid layouts as a delimited list within the string. For example, for tensor dims {5, 1, 10, 10} the memory layout string would be "|NCHW|NHWC|CHWN|" (plus any others). Notice that I've included delims at the start and end which clearly marks it as a multiple-layout string. If the first character is a delimiter we would require "|\<layout\>|" within the string for a match. We'd also add helper function(s) to handle all this conditional stuff such as:
```
TensorDescriptor::IsCompatibleLayout(const std::string& test_layout, const std::string& possible_layouts){...}
```

BrianHarrisonAMD commented:
> I guess they aren't going to supply more information in response to Evgenii's questions, so this LGTM unless you want to ping them again.
> 
> However, this has made it more clear how the string layouts are a limiting design flaw. What we need for more generic use cases such as this is a set of enums so the solvers can quickly determine whether the memory layout is appropriate for their kernels by checking flag(s). That of course would require a very large refactor. We could mimic it quite easily using the existing API though with little added compute expense, by simply including all valid layouts as a delimited list within the string. For example, for tensor dims {5, 1, 10, 10} the memory layout string would be "|NCHW|NHWC|CHWN|" (plus any others). Notice that I've included delims at the start and end which clearly marks it as a multiple-layout string. If the first character is a delimiter we would require "|<layout>|" within the string for a match. We'd also add helper function(s) to handle all this conditional stuff such as:
> 
> ```
> TensorDescriptor::IsCompatibleLayout(const std::string& test_layout, const std::string& possible_layouts){...}
> ```

Yea I agree.
I think I wrote something up, but then replaced it once I realized the test breaks were due to overwriting the calculated layout when it wasn't needed.

A full proper fix would account for the fact that multiple layouts might be supported, and we would need solvers to check if their required layout is one of them. This also also can impact the solvers where it supports multiple layouts, but one is faster. 

MISA & CK for example can support either, but if NHWC is allowed, then it should run that path instead.

For now I think this gives us a bit more flexibility, but there is still more room to improve if we have time.

BrianHarrisonAMD committed:
Fix test case not setting layout for both tensors

BrianHarrisonAMD committed:
Reviewed original regression issue, and added another case to ensure this behavior is captured

With the more relaxed tensors, it'll find a matching layout for all shapes and use that. If we have C == 1, then NCHW can be picked if not all tensors set the same layout.
